File: pufferlib/ocean/channelwave/channelwave.c
#include "channelwave.h"

// Helper to sample uniform ints
static inline int rand_int(int max_exclusive) {
    return rand() % max_exclusive;
}

static inline int8_t rand_sign() {
    return (rand() & 1) ? 1 : -1;
}

// Log episode completion
static void log_episode(ChannelWave* env, int agent_idx) {
    float ep_len = (float)env->episode_step[agent_idx];
    if (ep_len > 0) {
        env->log.score += env->episode_return[agent_idx];
        env->log.hearts_collected += env->hearts_collected[agent_idx];
        env->log.hearts_missed += env->hearts_missed[agent_idx];
        env->log.sync_accuracy += (float)env->sync_ticks[agent_idx] / ep_len;
        env->log.episode_length += ep_len;
        env->log.n += 1.0f;
    }
}

// Get instance index for an agent
static inline int get_instance(ChannelWave* env, int agent_idx) {
    return agent_idx / env->agents_per_instance;
}

// Get first channel index for an instance
static inline int get_instance_channel_start(ChannelWave* env, int instance) {
    return instance * env->channels_per_instance;
}

// Reset single agent state
static void reset_agent(ChannelWave* env, int agent_idx) {
    int instance = get_instance(env, agent_idx);
    int channel_start = get_instance_channel_start(env, instance);
    
    env->agent_pos[agent_idx] = 0;              // Start at Null (center)
    env->agent_channel[agent_idx] = channel_start;  // Start on first channel of instance
    env->move_cd[agent_idx] = 0;
    env->episode_return[agent_idx] = 0.0f;
    env->episode_step[agent_idx] = 0;
    env->hearts_collected[agent_idx] = 0;
    env->hearts_missed[agent_idx] = 0;
    env->sync_ticks[agent_idx] = 0;
}

// Per-channel spawn counter for deterministic alternation
static int spawn_count[32768] = {0};  // Track spawn count per channel

// Spawn reward on a channel based on spawn mode
static void spawn_reward(ChannelWave* env, int channel) {
    int8_t sign;
    
    if (env->spawn_mode == 0 || env->spawn_mode == 2) {
        // Deterministic alternating per channel
        // Each channel has its own alternation state based on spawn count
        sign = ((spawn_count[channel] % 2) == 0) ? 1 : -1;
        spawn_count[channel]++;
    } else {
        // Random sign (spawn_mode == 1)
        sign = rand_sign();
    }
    
    env->reward_pos[channel] = sign;
    env->reward_timer[channel] = env->channel_produce[channel];
    env->reward_claimed[channel] = 0;
}

// Reset environment state
void c_reset(ChannelWave* env) {
    // Reset all channels
    for (int c = 0; c < env->num_channels; c++) {
        env->reward_pos[c] = 0;
        env->reward_timer[c] = 0;
        env->reward_claimed[c] = 0;
        spawn_count[c] = 0;  // Reset per-channel spawn counter
        // Initialize spawn timers with phase offsets
        env->spawn_timer[c] = env->channel_phase[c];
    }
    
    env->step = 0;
    env->alternation_idx = 0;

    // Reset all agents
    for (int i = 0; i < env->num_agents; i++) {
        reset_agent(env, i);
    }

    // Initial spawn: all channels with phase=0 get a reward
    // This works for both single-channel (mode 0/1) and multi-channel (mode 2)
    for (int c = 0; c < env->num_channels; c++) {
        if (env->spawn_timer[c] == 0) {
            spawn_reward(env, c);
            env->spawn_timer[c] = env->channel_period[c];
        }
    }
}

// Single step
void c_step(ChannelWave* env) {
    const int agents = env->num_agents;
    const int channels = env->num_channels;

    // ===== AGENT MOVEMENT =====
    for (int i = 0; i < agents; i++) {
        // Get instance boundaries for this agent
        int instance = get_instance(env, i);
        int ch_start = get_instance_channel_start(env, instance);
        int ch_end = ch_start + env->channels_per_instance;
        
        // Parse action: [channel_offset, position_idx]
        // channel_offset is relative to instance's first channel
        int channel_offset = env->actions[2*i];
        if (channel_offset < 0) channel_offset = 0;
        if (channel_offset >= env->channels_per_instance) channel_offset = env->channels_per_instance - 1;
        int target_channel = ch_start + channel_offset;

        int pos_idx = env->actions[2*i + 1];
        if (pos_idx < 0) pos_idx = 0;
        if (pos_idx > 2) pos_idx = 2;
        int8_t target_pos = (int8_t)(pos_idx - 1); // map {0,1,2} -> {-1,0,1}

        // Movement with cooldown
        unsigned char moved = (target_pos != env->agent_pos[i]) || 
                              (target_channel != env->agent_channel[i]);
        
        if (env->move_cd[i] == 0 && moved) {
            env->agent_pos[i] = target_pos;
            env->agent_channel[i] = (int16_t)target_channel;
            env->move_cd[i] = env->t_m;
        } else if (env->move_cd[i] > 0) {
            env->move_cd[i]--;
        }

        // Apply movement penalty (step cost)
        env->rewards[i] = -env->move_penalty;
        env->episode_return[i] -= env->move_penalty;
    }

    // ===== REWARD COLLECTION & SYNC TRACKING =====
    for (int i = 0; i < agents; i++) {
        int ch = env->agent_channel[i];
        int8_t pos = env->agent_pos[i];
        int8_t reward_pos = env->reward_pos[ch];
        int16_t timer = env->reward_timer[ch];
        
        unsigned char reward_active = (timer > 0) && (reward_pos != 0);
        unsigned char at_reward = reward_active && (pos == reward_pos);
        unsigned char at_null = (pos == 0);

        // Track synchronization: agent is "synced" if at correct position
        if (at_reward) {
            env->sync_ticks[i]++;
        }

        // Reward collection
        if (env->reward_mode == 0) {
            // Discrete-once: reward only on first collection
            if (at_reward && !env->reward_claimed[ch]) {
                env->rewards[i] += env->heart_reward;
                env->episode_return[i] += env->heart_reward;
                env->reward_claimed[ch] = 1;
                env->hearts_collected[i]++;
            }
        } else {
            // Continuous: reward every tick at correct position
            if (at_reward) {
                env->rewards[i] += env->heart_reward;
                env->episode_return[i] += env->heart_reward;
                if (!env->reward_claimed[ch]) {
                    env->hearts_collected[i]++;
                    env->reward_claimed[ch] = 1;
                }
            }
        }

        // ===== OBSERVATIONS =====
        // [reward_sign_if_visible, reward_timer_if_visible, agent_pos, move_cd, channel_offset, spawn_timer]
        unsigned char visible = at_null;  // Only see reward location from Null
        int16_t obs_sign = visible ? reward_pos : 0;
        int16_t obs_time = visible ? timer : 0;
        int16_t obs_spawn = visible ? env->spawn_timer[ch] : 0;
        
        // Report channel as offset within instance (not global index)
        int instance = get_instance(env, i);
        int ch_start = get_instance_channel_start(env, instance);
        int ch_offset = ch - ch_start;

        int obs_base = i * 6;
        env->observations[obs_base + 0] = (unsigned char)(obs_sign + 128);
        env->observations[obs_base + 1] = (unsigned char)(obs_time > 255 ? 255 : (obs_time < 0 ? 0 : obs_time));
        env->observations[obs_base + 2] = (unsigned char)(pos + 1);
        env->observations[obs_base + 3] = (unsigned char)(env->move_cd[i] > 255 ? 255 : env->move_cd[i]);
        env->observations[obs_base + 4] = (unsigned char)ch_offset;
        env->observations[obs_base + 5] = (unsigned char)(obs_spawn > 255 ? 255 : (obs_spawn < 0 ? 0 : obs_spawn));
    }

    // ===== CHANNEL TIMER UPDATES =====
    for (int c = 0; c < channels; c++) {
        // Decay reward timer
        if (env->reward_timer[c] > 0) {
            env->reward_timer[c]--;
            if (env->reward_timer[c] == 0) {
                // Reward expired - count as missed if unclaimed
                if (!env->reward_claimed[c]) {
                    // Attribute miss to agents on this channel
                    for (int i = 0; i < agents; i++) {
                        if (env->agent_channel[i] == c) {
                            env->hearts_missed[i]++;
                        }
                    }
                }
                env->reward_pos[c] = 0;
                env->reward_claimed[c] = 0;
            }
        }

        // Spawn timer countdown
        if (env->spawn_timer[c] > 0) {
            env->spawn_timer[c]--;
        }
        
        // Spawn new reward when timer hits zero
        // All channels spawn independently (works for instance-based layout)
        if (env->spawn_timer[c] == 0 && env->reward_timer[c] == 0) {
            spawn_reward(env, c);
            env->spawn_timer[c] = env->channel_period[c];
        }
    }

    env->step++;

    // ===== EPISODE TERMINATION =====
    for (int i = 0; i < agents; i++) {
        env->episode_step[i]++;
        
        if (env->max_steps > 0 && env->episode_step[i] >= env->max_steps) {
            env->terminals[i] = 1;
            log_episode(env, i);
            reset_agent(env, i);
        } else {
            env->terminals[i] = 0;
        }
    }
}

void c_render(ChannelWave* env) {
    // No-op render stub
}

void c_close(ChannelWave* env) {
    free(env->channel_period);
    free(env->channel_produce);
    free(env->channel_phase);
    free(env->reward_pos);
    free(env->reward_timer);
    free(env->spawn_timer);
    free(env->reward_claimed);
    free(env->agent_pos);
    free(env->agent_channel);
    free(env->move_cd);
    free(env->episode_return);
    free(env->episode_step);
    free(env->hearts_collected);
    free(env->hearts_missed);
    free(env->sync_ticks);
    if (env->alternation_pattern) {
        free(env->alternation_pattern);
    }
}

File: pufferlib/ocean/channelwave/channelwave.py
"""
ChannelWave Environment for PufferLib

A high-performance environment for studying phase-locking and synchronization
dynamics in reinforcement learning, inspired by periodic reward experiments.

Key Concepts:
- N channels, each with +, -, and Null positions
- Rewards appear as square waves with configurable period and duty cycle
- Agents must learn to synchronize with reward patterns
- Null position provides observability (like being at center of hallway)

Spawn Modes:
- "deterministic" (0): Single channel with deterministic alternation pattern
- "random" (1): Single channel, random +/- sign each spawn
- "independent" (2): Each channel has independent timing (for Arnold Tongue experiments)

This environment maps to the "Periodic Hearts" experimental framework:
- channel_period = T (total period in ticks)
- channel_produce = T_produce (duty cycle high duration)
- channel_phase = φ (phase offset)
- t_m = d_h / r_a (movement time = critical period)
- max_steps = L (episode length)
"""

import gymnasium
import numpy as np

import pufferlib
from pufferlib.ocean.channelwave import binding


class ChannelWave(pufferlib.PufferEnv):
    """High-throughput channel signal game for phase-locking experiments.

    Args:
        num_channels: Number of independent channels (N)
        num_agents: Number of agents (M)
        max_steps: Episode length in ticks (L)
        
        # Timing parameters (support per-channel or scalar)
        default_period: Default T for all channels (ticks per cycle)
        default_produce: Default T_produce (ticks reward is available)
        channel_periods: Optional list of per-channel periods [T1, T2, ...]
        channel_produces: Optional list of per-channel produce times
        channel_phases: Optional list of per-channel phase offsets [φ1, φ2, ...]
        
        # Movement
        t_m: Movement cooldown (analogous to d_h/r_a, the critical period)
        move_penalty: Step cost R_S (typically 0 or small negative)
        
        # Rewards
        heart_reward: Reward per heart collected R_H
        reward_mode: "discrete" (once per heart) or "continuous" (every tick at heart)
        
        # Spawn mode
        spawn_mode: "deterministic", "random", or "independent"
        alternation_pattern: Custom pattern for deterministic mode, e.g. [1, -1, 1, -1]
    """

    def __init__(
        self,
        num_channels: int = 1,
        num_agents: int = 1,
        max_steps: int = 500,
        # Timing
        default_period: int = 40,
        default_produce: int = 20,
        channel_periods: list = None,
        channel_produces: list = None,
        channel_phases: list = None,
        # Movement
        t_m: int = 10,
        move_penalty: float = 0.0,
        # Rewards
        heart_reward: float = 1.0,
        reward_mode: str = "discrete",
        # Spawn mode
        spawn_mode: str = "deterministic",
        alternation_pattern: list = None,
        # PufferLib standard
        buf=None,
        seed: int = 0,
        render_mode: str = None,
    ):
        self.num_agents = num_agents
        self.single_observation_space = gymnasium.spaces.Box(
            low=0, high=255, shape=(6,), dtype=np.uint8
        )
        # Actions: (channel_idx, position_idx)
        # position_idx: 0=minus, 1=null, 2=plus
        self.single_action_space = gymnasium.spaces.MultiDiscrete(
            [num_channels, 3], dtype=np.int32
        )
        self.render_mode = render_mode

        super().__init__(buf)

        # Convert string modes to int
        reward_mode_int = 0 if reward_mode == "discrete" else 1
        spawn_mode_map = {"deterministic": 0, "random": 1, "independent": 2}
        spawn_mode_int = spawn_mode_map.get(spawn_mode, 0)

        # Store config for introspection
        self._config = {
            "num_channels": num_channels,
            "num_agents": num_agents,
            "max_steps": max_steps,
            "default_period": default_period,
            "default_produce": default_produce,
            "t_m": t_m,
            "spawn_mode": spawn_mode,
            "critical_period": t_m,  # T_c = t_m in this abstraction
        }

        self.c_envs = binding.vec_init(
            self.observations,
            self.actions,
            self.rewards,
            self.terminals,
            self.truncations,
            1,  # Single environment that handles all agents internally
            seed,
            num_channels=num_channels,
            num_agents=num_agents,
            max_steps=max_steps,
            default_period=default_period,
            default_produce=default_produce,
            channel_periods=list(channel_periods) if channel_periods else None,
            channel_produces=list(channel_produces) if channel_produces else None,
            channel_phases=list(channel_phases) if channel_phases else None,
            t_m=t_m,
            move_penalty=float(move_penalty),
            heart_reward=float(heart_reward),
            reward_mode=reward_mode_int,
            spawn_mode=spawn_mode_int,
            alternation_pattern=list(alternation_pattern) if alternation_pattern else None,
        )

    @property
    def config(self):
        """Return environment configuration for logging."""
        return self._config

    def reset(self, seed=0):
        binding.vec_reset(self.c_envs, seed)
        return self.observations, []

    def step(self, actions):
        self.actions[:] = actions
        binding.vec_step(self.c_envs)
        info = [binding.vec_log(self.c_envs)]
        return self.observations, self.rewards, self.terminals, self.truncations, info

    def render(self):
        pass

    def close(self):
        binding.vec_close(self.c_envs)


# ============================================================================
# Convenience constructors for specific experimental setups
# ============================================================================

def make_hallway(
    distance: int = 10,
    period: int = None,
    produce: int = None,
    duty_cycle: float = 0.5,
    num_agents: int = 1,
    max_steps: int = 500,
    **kwargs
):
    """Create a single-channel hallway experiment (primary experiment from doc).
    
    Args:
        distance: d_h - hallway distance (determines critical period T_c = distance)
        period: T - square wave period. If None, defaults to 2 * distance
        produce: T_produce - time heart is available. If None, uses duty_cycle
        duty_cycle: Fraction of period that heart is available (default 0.5)
        num_agents: Number of agents
        max_steps: Episode length L
    
    The critical period is T_c = distance (since r_a = 1 in this abstraction).
    """
    if period is None:
        period = 2 * distance  # Default: agent can just make it with perfect timing
    if produce is None:
        produce = int(period * duty_cycle)
    
    return ChannelWave(
        num_channels=1,
        num_agents=num_agents,
        max_steps=max_steps,
        default_period=period,
        default_produce=produce,
        t_m=distance,  # Critical period = movement time
        spawn_mode="deterministic",
        **kwargs
    )


def make_arnold_tongue_pair(
    f1: float,
    f2: float,
    phase_offset: int = None,
    distance: int = 10,
    duty_cycle: float = 0.5,
    num_agents: int = 1,
    max_steps: int = 500,
    **kwargs
):
    """Create a 2-channel experiment for Arnold Tongue analysis.
    
    Args:
        f1: Frequency of channel 0 (cycles per tick)
        f2: Frequency of channel 1 (cycles per tick)
        phase_offset: Phase offset for channel 1 (if None, anti-phase)
        distance: Movement time / critical period
        duty_cycle: Duty cycle for both channels
        
    Returns:
        ChannelWave with two independently-timed channels.
    """
    T1 = max(1, int(1.0 / f1))
    T2 = max(1, int(1.0 / f2))
    
    produce1 = max(1, int(T1 * duty_cycle))
    produce2 = max(1, int(T2 * duty_cycle))
    
    if phase_offset is None:
        phase_offset = T2 // 2  # Anti-phase by default
    
    return ChannelWave(
        num_channels=2,
        num_agents=num_agents,
        max_steps=max_steps,
        default_period=T1,
        default_produce=produce1,
        channel_periods=[T1, T2],
        channel_produces=[produce1, produce2],
        channel_phases=[0, phase_offset],
        t_m=distance,
        spawn_mode="independent",
        **kwargs
    )


def make_four_arm_cross(
    pattern: str = "alternating",  # or "paired"
    period: int = 40,
    distance: int = 10,
    num_agents: int = 2,
    max_steps: int = 500,
    **kwargs
):
    """Create a 4-channel cross experiment (multi-agent secondary experiment).
    
    Args:
        pattern: "alternating" = [H1,H2,H1,H2] or "paired" = [H1,H1,H2,H2]
        period: Cycle period for all channels
        distance: d_side (adjacent channel distance)
        num_agents: Number of competing/cooperating agents
    """
    produce = period // 2
    
    if pattern == "alternating":
        # [H1, H2, H1, H2] - adjacent channels have different "colors"
        phases = [0, period // 4, period // 2, 3 * period // 4]
    else:  # paired
        # [H1, H1, H2, H2] - adjacent channels have same "color"
        phases = [0, 0, period // 2, period // 2]
    
    return ChannelWave(
        num_channels=4,
        num_agents=num_agents,
        max_steps=max_steps,
        default_period=period,
        default_produce=produce,
        channel_phases=phases,
        t_m=distance,
        spawn_mode="independent",
        **kwargs
    )


# ============================================================================
# Performance test
# ============================================================================

if __name__ == "__main__":
    import time
    
    print("Testing basic ChannelWave...")
    N = 4096
    env = ChannelWave(num_channels=1, num_agents=N, max_steps=500, spawn_mode="deterministic")
    env.reset()
    
    CACHE = 1024
    actions = np.zeros((CACHE, N, 2), dtype=np.int32)
    actions[..., 0] = 0  # Always channel 0
    actions[..., 1] = np.random.randint(0, 3, size=(CACHE, N))  # Random position
    
    steps = 0
    start = time.time()
    while time.time() - start < 5:
        env.step(actions[steps % CACHE])
        steps += 1
    
    sps = int(env.num_agents * steps / (time.time() - start))
    print(f"ChannelWave SPS: {sps:,}")
    env.close()
    
    print("\nTesting hallway experiment...")
    env = make_hallway(distance=10, period=40, num_agents=1, max_steps=100)
    obs, _ = env.reset()
    print(f"Config: {env.config}")
    print(f"Observation shape: {obs.shape}")
    
    total_reward = 0
    for _ in range(100):
        # Simple policy: alternate between + and -
        action = np.array([[0, 2]])  # Stay at + position
        obs, rewards, terminals, truncations, info = env.step(action)
        total_reward += rewards[0]
        if terminals[0]:
            print(f"Episode ended. Total reward: {total_reward:.2f}")
            print(f"Info: {info}")
            break
    
    env.close()
    print("Tests passed!")
